### ISSUE TITLE
Support for SSH cloning

### ISSUE THREAD

**Issue by sluongng** [esrlabs/josh#22]
Please add support for SSH protocol 👍

**Comment by jonatansberg** [comment]
We're looking in to using Josh in order to open source parts of a bigger trunk based project at my company. This includes mirroring subtrees to public repos and using a subtree approach to compose projects that would depend on one or more of our upstream repositories.

Supporting SSH and public key auth would be crucial for this to work at a larger scale. I really don't want us to have to deal with everyones private git/github credentials.

I'm fairly new to this project (and rust in general) and only have a very limited understanding of what's what, but if there is any way we can contribute to making this happen I'd be happy for us to help!

**Comment by initcrash** [comment]
Hi, @jonatansberg 
it sounds like your use case, synchronizing parts of a bigger repo and re-composing projects from multiple dependencies, is very much like what we are using Josh for at ESR Labs.

Regarding ssh support:
A long time ago, before this thing even had a name, I implemented support for ssh but realized quickly that is
not suitable for making a proxy, so I switched over to http and removed ssh support.

The problem is that in order to use ssh Josh would need an additional pair of keys per user.
Those keys would either have to be maintained manually, or Josh would need extra support for fetching/installing keys in the upstream server, for which, once again it would need credentials. So while ssh looks simpler at first, setting up a proxy with it is actually much more complex than with http where credentials can simply be forwarded.

If you like we can have a chat/call about this in the next days. I'm curious about your use case and why http won't work for you.
Also maybe I'm missing something and there is good solution to managing ssh keys, in which case I will gladly re-add ssh protocol support to Josh.

**Comment by fire** [comment]
I use ssh via hardware keys. Could not evaluate josh.

**Comment by LMG** [comment]
This is work in progress using SSH agent fowarding: https://github.com/josh-project/josh/commit/7028451ee1242485233cf1efebcdc1a47a7e4c46

**Comment by RalfJung** [comment]
Interestingly, ssh pushing already works when using an https URL and `pushInsteadOf`. That makes me wonder whether `insteadOf` could be used to also make josh pull via ssh.

Or is this issue here less about the SSH protocol support and more about how to get josh to be able to access the necessary SSH keys?

**Comment by christian-schilling** [comment]
What do you mean "already works"? How can it?

But yes, the issue with SSH is entirely about josh being able to access the keys somehow.

**Comment by RalfJung** [comment]
> What do you mean "already works"? How can it?

I am setting `--remote=https://github.com` and my .gitconfig contains
```
[url "git@github.com:"]
        pushInsteadOf = https://github.com/
```
This makes josh push via SSH just fine.

**Comment by LMG** [comment]
This doesn't, though: according to the doc, "Any URL that starts with this value will not be pushed to; instead, it will be rewritten to start with <base>, and the resulting URL will be pushed to."

So it just pushes through Https.

**Comment by RalfJung** [comment]
> So it just pushes through Https.

No, that wouldn't work, since it doesn't know my GH login. I have been using this config since ~forever, it is quite convenient to pull via https without authentication and then use SSH and key-based authentication for pushing.

That config means that if the to-be-pushed URL starts with `https://github.com/`, that prefix is replaced by `git@github.com:`, which is the interpreted as an SSH URL and handled by then regular git pushing logic.

I set this locally on my machine.

> If it just gets them from its environment, josh can already do it I believe, directly with an ssh url

Sure, using an SSH URL would also work.
